pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.pinot.core.data.readers.MultiplePinotSegmentRecordReader;
+import org.apache.pinot.core.indexsegment.IndexSegment;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+
+
+/**
+ * Dimension Table is a special type of OFFLINE table which is assigned to all servers
+ * and is used to execute a LOOKUP Transform Function. They should be small enough to fit in memory(<100MB).
+ *
+ * DimensionTableDataManager uses Registry of Singletons pattern to store one instance per table
+ * which can be accessed via 'getInstanceByTableName' static method.
+ */
+public class DimensionTableDataManager extends OfflineTableDataManager {
+  // Store singletons per table in this map
+  private static final Map<String, DimensionTableDataManager> _instances = new HashMap<>();
+
+  public static DimensionTableDataManager getInstanceByTableName(String tableName) {
+    if (!_instances.containsKey(tableName)) {
+      _instances.put(tableName, new DimensionTableDataManager(tableName));
+    }
+    return _instances.get(tableName);
+  }
+
+  // DimensionTableDataManager Instance
+
+  private String _tableName;
+  // _lookupTable is a HashMap used for fetching records from a table given the primary key
+  // TODO needs to be periodically recycled
+  private final Map<PrimaryKey, GenericRow> _lookupTable = new HashMap<>();
+  private Boolean _lookupTableLoaded = false;
+  private final ReadWriteLock _rwl = new ReentrantReadWriteLock();
+  private final Lock _lookupTableReadLock = _rwl.readLock();
+  private final Lock _lookupTableWriteLock = _rwl.writeLock();
+
+  private DimensionTableDataManager(String tableName) {
+    _tableName = tableName;
+  }
+
+  private void prepareLookupTable () throws Exception {
+    List<SegmentDataManager> segmentManagers = acquireAllSegments();
+    List<File> indexDirs = new ArrayList<>();
+    List<String> primaryKeyColumns = new ArrayList<>();
+
+    for (SegmentDataManager segmentManager: segmentManagers) {
+      IndexSegment indexSegment = segmentManager.getSegment();
+      indexDirs.add(indexSegment.getSegmentMetadata().getIndexDir());
+      primaryKeyColumns = indexSegment.getSegmentMetadata().getSchema().getPrimaryKeyColumns();
+    }
+
+    // TODO validate primary key columns exist
+    if (primaryKeyColumns == null) {
+      primaryKeyColumns = new ArrayList<>(Arrays.asList("teamID"));
+    }
+
+    _lookupTableWriteLock.lock();
+    try {
+      MultiplePinotSegmentRecordReader reader = new MultiplePinotSegmentRecordReader(indexDirs);
+      while (reader.hasNext()) {
+        GenericRow row = reader.next();
+        _lookupTable.put(row.getPrimaryKey(primaryKeyColumns), row);
+      }
+      _lookupTableLoaded = true;
+    } finally {
+      _lookupTableWriteLock.unlock();
+    }
+  }
+
+  public GenericRow lookupRowByPrimaryKey(PrimaryKey pk) throws Exception {
+    if (!_lookupTableLoaded) {
+      prepareLookupTable();
```

Comment:
Loading at query time could cause high latency. Maybe we can always pre-load the dimension table

Reply:
Good point, we can do that. However there is also an argument for lazy-loading. Depending on the number and size of tables loaded, this can prevent unnecessary memory usage in cases where a dim table isn't accessed frequently. Also, if the table is already mmap'ed, loading it into a HashMap should be relatively fast.

In my test case, I filled the 'baseballTeams' table with 1MM rows, compressed segment size being ~25MB. Loading time (for the HashMap) seems to float around 1.8s on my laptop, which is not terrible. Let me know what you think.

Reply:
Update: I ended up going with your original recommendation, by overwriting the `addSegment` method, which also nicely solves the recycling issue when a new segment is added. 👍